### PR TITLE
Enable assignment & visualisation with partial query graphs

### PR DESCRIPTION
**Pull request by nickjcroucher**
To enable full database querying with beebop:
- This workflow saves a partial query graph after assignment, along with a list of the sequences in the network components containing queries
- This enables visusaliation of only the clusters that contain queries, rather than the full database
- A flag is added to enable distances to be recalculated for tree generation, rather than read from the file (question: would it be useful to have an option in which we sparsified the pairwise distance matrix to only include within-cluster distances?)
- I update the CI tests to allow testing of the beebop workflow
- This revealed a problem with the lineage assignment C++ code, which I think should now be resolved

### CODE REVIEW

**Comment by johnlees** [comment]
@nickjcroucher is this ready for review? I feel we should release and fix the conda builds as the dict bug keeps coming up

File: PopPUNK/visualise.py [hunk 1]
```diff
@@ -84,12 +84,19 @@ def get_options():
...
+                        action = 'store_true')
     iGroup.add_argument('--network-file',
                         help='Specify a file to use for any graph visualisations',
                         type = str)
     iGroup.add_argument('--display-cluster',
                         help='Column of clustering CSV to use for plotting',
                         default=None)
+    iGroup.add_argument('--use-partial-query-graph',
+                        help='File listing sequences in partial query graph after assignment',
+                        default=None)
```
Comment (johnlees) [inline PopPUNK/visualise.py:99]: Similar note to arg above

File: test/test-update.py
Comment (johnlees) [inline test/test-update.py:1]: This file needs whitespace properly removed (lots of spaces at starts of sentences)

**Review by johnlees — CHANGES_REQUESTED** [review]

File: PopPUNK/__init__.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 
 '''PopPUNK (POPulation Partitioning Using Nucleotide Kmers)'''
 
-__version__ = '2.7.2'
+__version__ = '2.7.3'
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/__init__.py]: Done in 118861b.

File: PopPUNK/network.py [hunk 3]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+    if use_gpu:
+        sys.stderr.write('Not compatible with GPU networks yet\n')
+        query_subgraph = G
+    else:
+        # Identify network components containing queries
+        component_dict = gt.label_components(G)[0]
+        components_with_query = set()
+        for i in range(num_refs,G.num_vertices()):
+          v = G.vertex(i)  # Access vertex by index
+          components_with_query.add(component_dict[v])
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: Fixed in 16551de.

File: PopPUNK/network.py [hunk 4]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+        for n, vertex in enumerate(G.vertices()):
+            if reflist[n] in samples_to_keep_set:
+                reference_vertex[vertex] = True
+            else:
+                reference_vertex[vertex] = False
+        G_new = gt.GraphView(G, vfilt = reference_vertex)
+        G_new = gt.Graph(G_new, prune = True)
+    return G_new
+
+def retain_only_query_clusters(G, rlist, qlist, use_gpu = False):
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: Done in 808f0be.

File: PopPUNK/network.py [hunk 5]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+        pruned_names (list)
+            The labels of the sequences in the pruned network
+    """
+    num_refs = len(rlist)
+    components_with_query = []
+    combined_names = rlist + qlist
+    pruned_names = []
+    if use_gpu:
+        sys.stderr.write('Not compatible with GPU networks yet\n')
+        query_subgraph = G
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: Good spot, fixed in 808f0be.

File: PopPUNK/network.py [hunk 6]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+        use_gpu (bool)
+            Whether to use GPUs for network construction
+
+    Returns:
+        G (graph)
+            The resulting network
+        pruned_names (list)
+            The labels of the sequences in the pruned network
+    """
+    num_refs = len(rlist)
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: This was my way of clarifying the potential confusion highlighted below, removed in 808f0be.

File: PopPUNK/network.py [hunk 7]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+    combined_names = rlist + qlist
+    pruned_names = []
+    if use_gpu:
+        sys.stderr.write('Not compatible with GPU networks yet\n')
+        query_subgraph = G
+    else:
+        # Identify network components containing queries
+        component_dict = gt.label_components(G)[0]
+        components_with_query = set()
+        for i in range(num_refs,G.num_vertices()):
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: Vertices can only be referenced by integer, they are unlabelled to be compatible across GPU and non-GPU processing. The order is maintained with querying. Now clarified in a comment in 808f0be.

File: PopPUNK/plot.py [hunk 8]
```diff
@@ -553,7 +555,9 @@ def outputsForCytoscape(G, G_mst, isolate_names, clustering, outPrefix, epiCsv,
         suffix = '_cytoscape'
     else:
         suffix = suffix + '_cytoscape'
-    save_network(G, prefix = outPrefix, suffix = suffix, use_graphml = True)
+    if use_partial_query_graph is None:
+        sys.stderr.write('In here\n')
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/plot.py:559]: Well spotted, gone in 8f28b0f.

File: PopPUNK/network.py [hunk 9]
```diff
@@ -1920,32 +1920,96 @@ def prune_graph(prefix, reflist, samples_to_keep, output_db_name, threads, use_g
...
+            Whether to use GPUs for network construction
+
+    Returns:
+        G (graph)
+            The resulting network
+        pruned_names (list)
+            The labels of the sequences in the pruned network
+    """
+    num_refs = len(rlist)
+    components_with_query = []
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py:1992]: Gone in https://github.com/bacpop/PopPUNK/commit/808f0be45541523e9a2a6e83a630696eb94c1c83.

File: src/extend.cpp [hunk 10]
```diff
@@ -103,14 +103,17 @@ sparse_coo extend(const sparse_coo &sparse_rr_mat,
...
         if (i < nr_samples) {
           j = std::get<1>(sparse_rr_mat)[row_start_idx[i] + *rr_it];
         } else {
           j = *rr_it;
         }
         dist = rr_dists[*rr_it];
         ++rr_it;
+      } else {
+        printf("Insufficient distances for specified kNN value; try reducing the maximum search depth");
+        break;
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply src/extend.cpp]: Good point fixed in 0b94b2d.

File: PopPUNK/sketchlib.py [hunk 11]
```diff
@@ -226,10 +226,19 @@ def joinDBs(db1, db2, output, update_random = None):
...
-    db2_name = db2 + "/" + os.path.basename(db2) + ".h5"
+    
+    if not full_names:
+        join_prefix = output + "/" + os.path.basename(output)
+        db1_name = db1 + "/" + os.path.basename(db1) + ".h5"
+        db2_name = db2 + "/" + os.path.basename(db2) + ".h5"
+    else:
+        db1_name = db1
+        db2_name = db2
+        join_prefix = output
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/sketchlib.py:241]: This is consistent with the behaviour of `removeFromDB`, I liked removing the possibility of loading the wrong file, particularly if hiearchically structured (e.g. lineage models within strains).

File: PopPUNK/models.py [hunk 12]
```diff
@@ -1124,7 +1124,7 @@ def __init__(self, outPrefix, ranks, max_search_depth, reciprocal_only, count_un
         ClusterFit.__init__(self, outPrefix)
         self.type = 'lineage'
         self.preprocess = False
-        self.max_search_depth = max_search_depth
+        self.max_search_depth = max_search_depth+5
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/models.py]: Done in 6d32dcc.

File: test/test-update.py
Comment (nickjcroucher) [reply test/test-update.py:1]: I think these were added to make indentation consistent across the file (moved a comment in 727776d).

File: PopPUNK/visualise.py (hunk 1 above)
Comment (nickjcroucher) [reply PopPUNK/visualise.py:99]: I think this is helpfully consistent with the `--save-partial-query-graph` of `assign.py`.

File: PopPUNK/assign.py [hunk 13]
```diff
@@ -106,6 +107,9 @@ def get_options():
     queryingGroup.add_argument('--accessory', help='(with a \'refine\' or \'lineage\' model) '
                                                         'Use an accessory-distance only model for assigning queries '
                                                         '[default = False]', default=False, action='store_true')
+    queryingGroup.add_argument('--use-full-network', help='Use full network rather than reference network for querying [default = False]',
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/assign.py:110]: `use` feels more helpful than `force`.

**Review by johnlees — APPROVED** [review]

**Pull request merged**